### PR TITLE
`azurerm_data_factory_linked_service_data_lake_storage_gen2` - Supports managed identity auth through `use_managed_identity `

### LINKED ISSUE #6501: Support for adding Managed Identity to Linked Services to ADLS Gen 2 for Azure Data Factory.
## Managed Identity for Linked Service to ADLS Gen 2 for Azure Data Factory

As of January 2020, Azure Data Factory (ADF) now supports Managed Identity (formerly known as  Managed Service Identity - MSI) to connect to other Azure resources like Azure Data Lake Storage (ADLS).  Link to the update can be found [here](https://docs.microsoft.com/en-us/azure/data-factory/data-factory-service-identity)

It would be super nice, if we can perform this function in Terraform and add the corresponding role to the resource as a one step process.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### New or Affected Resource(s)

* azurerm_data_factory_linked_service_data_lake_storage_gen2

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
```

### References

* #0000

### PR DESCRIPTION
**Pull request by dnlbunting**
Closes  #6501

Adds a use_managed_identity flag to azurerm_data_factory_linked_service_data_lake_storage_gen2.

If both use_managed_identity=true and service_principal_key or service_principal_id are specified raises an error.

First contribution so feel free to add feedback

### CODE REVIEW

File: azurerm/internal/services/datafactory/data_factory_linked_service_data_lake_storage_gen2_resource.go [hunk 1]
```diff
@@ -58,21 +58,27 @@ func resourceArmDataFactoryLinkedServiceDataLakeStorageGen2() *schema.Resource {
 				ValidateFunc: validation.IsURLWithHTTPS,
 			},
 
+			"use_managed_identity": {
+				Type:     schema.TypeBool,
+				Optional: true,
```
Comment (mbfrahry) [inline azurerm/internal/services/datafactory/data_factory_linked_service_data_lake_storage_gen2_resource.go:63]: We'll want to add `ConflictsWith` here too to make sure we don't specify it with the other service principal attributes.

Comment (mbfrahry) [reply azurerm/internal/services/datafactory/data_factory_linked_service_data_lake_storage_gen2_resource.go:63]: And to round it all out we should throw in `AtLeastOneOf` with all of these attributes to make sure at least one of them gets specified when creating this resource.

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hey @dnlbunting, thanks for opening this PR! Things look nearly there but some of the validation you wrote can be better surfaced at the plan level so we error before Terraform even applies. I've detailed some of what you need below and feel free to reach out if you have any questions around it.

**Review by mbfrahry — COMMENTED** [review]
Actually, would it make sense to just infer that we should use managed identity if we don't specify the service principal? Or is it better for the user to purposefully specify that they want managed identity?

**Comment by dnlbunting** [comment]
@mbfrahry Thanks for the pointers on the schema validation, I'll make those changes. IMO making the managed identity opt-in rather the the default was clearer, since the managed identity has to be explicitly enabled in the parent ADF resource too.

**Comment by mbfrahry** [comment]
That works for me! I'll look forward to the schema validation changes

**Review by mbfrahry — APPROVED** [review]
LGTM!

**Comment by mbfrahry** [comment]
Thanks for this @dnlbunting. It should make it into the next release!

**Pull request merged**